Meta/CMake/Superbuild/CMakeLists.txt

```diff
@@ -0,0 +1,79 @@
+cmake_minimum_required(VERSION 3.16)
+
+# Variables that we really don't want the user to pass to the serenity cross-build because they will screw things up :^)
+set(LAGOM_ONLY_VARS "CMAKE_CXX_COMPILER;CMAKE_C_COMPILER")
+set(EXCLUDE_FROM_CL_VARS "SERENITY_ARCH;SERENITY_TOOLCHAIN")
+
+# https://stackoverflow.com/a/48555098
+# https://cmake.org/pipermail/cmake/2018-January/067002.html
+# MUST be done before call to 'project'
+get_cmake_property(vars CACHE_VARIABLES)
+foreach(var IN LISTS vars)
+    get_property(currentHelpString CACHE "${var}" PROPERTY HELPSTRING)
+    if("${currentHelpString}" MATCHES "No help, variable specified on the command line." OR "${currentHelpString}" STREQUAL "")
+        message("${var} = [${${var}}]  --  ${currentHelpString}") # uncomment to see the variables being processed
+        if (var IN_LIST EXCLUDE_FROM_CL_VARS)
+            continue()
+        elseif(var IN_LIST LAGOM_ONLY_VARS)
+            list(APPEND LAGOM_CL_ARGS "-D${var}=${${var}}")
+        else()
+            list(APPEND CL_ARGS "-D${var}=${${var}}")
+        endif()
+    endif()
+endforeach()
+
+list(APPEND LAGOM_CL_ARGS ${CL_ARGS})
+
+project(
+    SerenitySuperbuild
+    DESCRIPTION "Orchestrate host and target builds in a single build"
+    LANGUAGES NONE
+)
+
+get_filename_component(
+    SERENITY_SOURCE_DIR "${PROJECT_SOURCE_DIR}/../../.."
+    ABSOLUTE CACHE
+)
+set(SERENITY_ARCH "i686" CACHE STRING "Architecture to build for (i686 or x86_64)")
+set(SERENITY_TOOLCHAIN "GNU" CACHE STRING "Compliler toolchain to use for Serenity (GNU or Clang)")
+set(SERENITY_BUILD_DIR "${PROJECT_BINARY_DIR}/../${SERENITY_ARCH}")
+
+include(ExternalProject)
+
+ExternalProject_Add(
+    lagom
+    SOURCE_DIR "${SERENITY_SOURCE_DIR}/Meta/Lagom"
+    BINARY_DIR "${PROJECT_BINARY_DIR}/../lagom"
+    INSTALL_DIR "${PROJECT_BINARY_DIR}/../lagom-install"
+    CMAKE_GENERATOR Ninja
+    EXCLUDE_FROM_ALL YES
+    CMAKE_ARGS ${LAGOM_CL_ARGS}
+    CMAKE_CACHE_DEFAULT_ARGS
+    "-DCMAKE_INSTALL_PREFIX:PATH=<INSTALL_DIR>"
+    "-DUNICODE_CACHE_LOCATION:PATH=${SERENITY_BUILD_DIR}"
+    "-DBUILD_LAGOM:BOOL=${BUILD_LAGOM}"
+    # Always call the build step of tools, so keeping things up-to-date is easy
+    BUILD_ALWAYS YES
+    # Expose install step as a target, so it can be depended on
+    STEP_TARGETS install
+)
+
+ExternalProject_Add(
+    serenity
+    SOURCE_DIR "${SERENITY_SOURCE_DIR}"
+    BINARY_DIR "${SERENITY_BUILD_DIR}"
+    CMAKE_GENERATOR Ninja
+    CMAKE_ARGS ${CL_ARGS}
+    CMAKE_CACHE_DEFAULT_ARGS
+    # Tell the find_package(Lagom REQUIRED) command call where to find
+    # the CMake package
+    "-DCMAKE_PREFIX_PATH:PATH=${PROJECT_BINARY_DIR}/../lagom-install"
+    "-DUNICODE_CACHE_LOCATION:PATH=${SERENITY_BUILD_DIR}"
+    "-DSERENITY_ARCH:STRING=${SERENITY_ARCH}"
+    "-DSERENITY_TOOLCHAIN:STRING=${SERENITY_TOOLCHAIN}"
```

Comment:
[cmake-toolchains(7)](https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html) file should be used instead. Any reason why that is not so?

Reply:
I am autogenerating the Toolchain files so that we don't have to set environment variables all over the place just to use it. The previous iteration of the Toolchain file was just for Ports/ and required setting both $SERENITY_ARCH and $SERENITY_SOURCE_DIR in order to do anything, which is tedious for an incremental build.

I'm generating it in the Serenity CMakeLists instead of in the superbuild because I wasn't sure how proscriptive I wanted to be about using the superbuild, and I hadn't looked at how to change Meta/serenity.sh to play nicely with it yet.

Most of that comes from not knowing the best way to tool up the superbuild. For example, I looked at creating cmake-presets, but it didn't look like IDEs of today would pick that up and see the host and target sub builds properly :/. Having a pre-configure step of, run CMake in the super build directory in your terminal before telling CLion/VsCode to use these exact settings for configuring the target build seems... Less than ideal? I hope I'm just missing something.

Reply:
I will have to figure out a way to get my hands dirty with SerenityOS directly, but I still haven't switched to Linux host OS for pragmatic purposes on my laptop and the build tooling isn't exactly cross-platform friendly :^)

Yeah, you can forget about editors and IDEs directly recognizing superbuilds, it's not a common enough pattern that any of them bother with it. Consuming already generated child builds however is no problem, since they're just regular binary directories. Presets are "too new" and the existing tooling for those in editors and IDEs is terribad (of the ones I tried).